{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Single-cell Data Preparation and Preprocessing #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lyx/.local/lib/python3.9/site-packages/networkx/utils/backends.py:135: RuntimeWarning: networkx backend defined more than once: nx-loopback\n",
      "  backends.update(_get_backends(\"networkx.backends\"))\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import STED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = './demo/PBMC'\n",
    "out_dir = './demo/PBMC/test_CorEx'\n",
    "sc_count_file = os.path.join(data_dir,\"10k_PBMC_Multiome_RNA_counts.h5ad\")\n",
    "sc_anno_file = os.path.join(data_dir,\"MainCelltype.txt\")\n",
    "\n",
    "seed = 1\n",
    "ntopics = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(out_dir):\n",
    "        os.makedirs(out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create an scPreProcessing object to preprocess single-cell data. The sc_count_file parameter accepts both scRNA-Seq and single-cell epigenomic datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "scp = scPreProcessing()\n",
    "scp.set_data(sc_count_file=sc_count_file,sc_anno_file=sc_anno_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "scp.cell_selection()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The **gene_use** parameter defines the gene selection strategy. Options include 'All', 'MARKER', and 'HVG' (default is 'HVG'). <br>\n",
    "<br>\n",
    "When using the 'MARKER' method, the number of top marker genes is specified by ntopMarker (default: 200). For the 'HVG' method, the number of highly variable genes is controlled by ntopHvg (default: 500). <br>\n",
    "<br>\n",
    "If a **save_path** is provided, the list of selected genes will be saved to that directory. Additionally, gene_use accepts a list object to directly specify custom genes for downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# scp.geneWords(sc_anno_file, gene_use='HVG',save_path=out_dir)\n",
    "gene_use = pd.read_csv(os.path.join(data_dir,\"use_gene.csv\"),index_col=0)['0'].to_list()\n",
    "scp.geneWords(sc_anno_file, gene_use=gene_use)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Topic Extraction with scTopic #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameter Configuration for Data Preprocessing using the STED-CorEx Model\n",
    "scp.preprocessing(linear2log=True)\n",
    "\n",
    "# Parameter Configuration for Data Preprocessing using the STED-LDA Model\n",
    "# scp.preprocessing(linear2log=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**anchored_genes** supports nested lists, where each sublist contains the anchor genes for a specific topic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# anchor_words = []# Unsupervised\n",
    "anchor_words = [\n",
    "    [\"CD79A\", \"MS4A1\"],\n",
    "    [\"GNLY\", \"NKG7\"],\n",
    "    [\"CCL5\", \"GZMK\"],\n",
    "   [\"IL7R\", \"RORA\",\"TCF7\", \"LEF1\"],\n",
    "   [\"FCGR3A\", \"MS4A7\", \"LYZ\", \"CD14\"],\n",
    "   [\"FCER1A\", \"CST3\"]\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- reflect genes in expression ---\n",
      "6 anchored topics\n",
      "16 genes were registered\n",
      "--- extract cell type specific anchors ---\n",
      "6 cells\n",
      "set() type of anchored genes were removed (no specific after removing overlap)\n",
      "16 genes were registered\n",
      "seed number: 16\n",
      "seed_k: 16\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "scp.preForGuide(anchored_genes=anchor_words)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Construct a scTopic object for model training."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "sct = scTopic()\n",
    "sct.setData(scp,out_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configure the **hierarchy_topic** parameter to guide the hierarchical modeling process.<br> \n",
    "<br>\n",
    "For instance, when ntopics is set to 10 and hierarchy_topic is [6,2], the first CorEx iteration generates 10 topics, the second iteration builds upon these 10 topics to generate 6 new topics, and the third iteration further constructs 2 broader topics based on the 6 topics. <br>\n",
    "<br>\n",
    "When **tc_plot** is set to True, the resulting plots will be saved in the out_dir/model/ directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selecting the Anchored CorEx model.\n",
      "WARNING: Some words never appear (or always appear)\n",
      "WARNING: Some words never appear (or always appear)\n",
      "weight threshold is 0.000000 for graph with max of 72.000000 edges \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n"
     ]
    }
   ],
   "source": [
    "sct.trainCorEx(seed=seed,n_iter=200,n_ensemble=0,ntopics=ntopics,\n",
    "               anchor_strength=2,hierarchy_topic = [6,2],tc_plot=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively, hierarchical clustering is used to visualize the topic relationships obtained by the STED-CorEx model.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Some words never appear (or always appear)\n",
      "weight threshold is 0.000000 for graph with max of 72.000000 edges \n"
     ]
    }
   ],
   "source": [
    "ht = sct.model.hierarchy(n_hidden_list=[6,2],max_edges=200,plot =True,figfile=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = sct.model.plot_hierarchy(ht)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg class=\"main-svg\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"1000\" height=\"350\" style=\"\" viewBox=\"0 0 1000 350\"><rect x=\"0\" y=\"0\" width=\"1000\" height=\"350\" style=\"fill: rgb(255, 255, 255); fill-opacity: 1;\"/><defs id=\"defs-ad8c6f\"><g class=\"clips\"><clipPath id=\"clipad8c6fxyplot\" class=\"plotclip\"><rect width=\"743\" height=\"170\"/></clipPath><clipPath class=\"axesclip\" id=\"clipad8c6fx\"><rect x=\"177\" y=\"0\" width=\"743\" height=\"350\"/></clipPath><clipPath class=\"axesclip\" id=\"clipad8c6fy\"><rect x=\"0\" y=\"100\" width=\"1000\" height=\"170\"/></clipPath><clipPath class=\"axesclip\" id=\"clipad8c6fxy\"><rect x=\"177\" y=\"100\" width=\"743\" height=\"170\"/></clipPath></g><g class=\"gradients\"/><g class=\"patterns\"/></defs><g class=\"bglayer\"><rect class=\"bg\" x=\"177\" y=\"100\" width=\"743\" height=\"170\" style=\"fill: rgb(236, 239, 241); fill-opacity: 1; stroke-width: 0;\"/></g><g class=\"layer-below\"><g class=\"imagelayer\"/><g class=\"shapelayer\"/></g><g class=\"cartesianlayer\"><g class=\"subplot xy\"><g class=\"layer-subplot\"><g class=\"shapelayer\"/><g class=\"imagelayer\"/></g><g class=\"minor-gridlayer\"><g class=\"x\"/><g class=\"y\"/></g><g class=\"gridlayer\"><g class=\"x\"/><g class=\"y\"/></g><g class=\"zerolinelayer\"/><g class=\"layer-between\"><g class=\"shapelayer\"/><g class=\"imagelayer\"/></g><path class=\"xlines-below\"/><path class=\"ylines-below\"/><g class=\"overlines-below\"/><g class=\"xaxislayer-below\"/><g class=\"yaxislayer-below\"/><g class=\"overaxes-below\"/><g class=\"overplot\"><g class=\"xy\" transform=\"translate(177,100)\" clip-path=\"url(#clipad8c6fxyplot)\"><g class=\"scatterlayer mlayer\"><g class=\"trace scatter tracee2192f\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M0,76.5L243.72,76.5L243.72,93.5L0,93.5\" style=\"vector-effect: none; fill: none; stroke: rgb(61, 153, 112); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter traced2b57b\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M243.72,85L298.9,85L298.9,110.5L0,110.5\" style=\"vector-effect: none; fill: none; stroke: rgb(0, 116, 217); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter tracee0d944\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M0,59.5L341.69,59.5L341.69,97.75L298.9,97.75\" style=\"vector-effect: none; fill: none; stroke: rgb(0, 116, 217); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter traceea3414\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M341.69,78.63L352.2,78.63L352.2,127.5L0,127.5\" style=\"vector-effect: none; fill: none; stroke: rgb(0, 116, 217); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter trace31348f\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M0,42.5L452.48,42.5L452.48,103.06L352.2,103.06\" style=\"vector-effect: none; fill: none; stroke: rgb(0, 116, 217); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter trace3cb63f\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M452.48,72.78L515.98,72.78L515.98,144.5L0,144.5\" style=\"vector-effect: none; fill: none; stroke: rgb(0, 116, 217); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter tracee54999\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M0,25.5L567.08,25.5L567.08,108.64L515.98,108.64\" style=\"vector-effect: none; fill: none; stroke: rgb(0, 116, 217); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter trace7b4893\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M567.08,67.07L725.84,67.07L725.84,161.5L0,161.5\" style=\"vector-effect: none; fill: none; stroke: rgb(0, 116, 217); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter tracefbcea8\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M0,8.5L743,8.5L743,114.29L725.84,114.29\" style=\"vector-effect: none; fill: none; stroke: rgb(0, 116, 217); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g></g></g></g><path class=\"xlines-above crisp\" d=\"M176,270.5H921M176,99.5H921\" style=\"fill: none; stroke-width: 1px; stroke: rgb(235, 240, 248); stroke-opacity: 1;\"/><path class=\"ylines-above crisp\" d=\"M176.5,100V270M920.5,100V270\" style=\"fill: none; stroke-width: 1px; stroke: rgb(235, 240, 248); stroke-opacity: 1;\"/><g class=\"overlines-above\"/><g class=\"xaxislayer-above\"><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(177,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(311.03999999999996,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(445.09,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(579.13,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(713.17,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(847.21,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" transform=\"translate(177,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\">0</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(311.03999999999996,0)\">0.5</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(445.09,0)\">1</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(579.13,0)\">1.5</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(713.17,0)\">2</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(847.21,0)\">2.5</text></g></g><g class=\"yaxislayer-above\"><path class=\"ytick ticks crisp\" d=\"M176,0h-5M921,0h5\" transform=\"translate(0,108.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M176,0h-5M921,0h5\" transform=\"translate(0,125.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M176,0h-5M921,0h5\" transform=\"translate(0,142.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M176,0h-5M921,0h5\" transform=\"translate(0,159.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M176,0h-5M921,0h5\" transform=\"translate(0,176.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M176,0h-5M921,0h5\" transform=\"translate(0,193.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M176,0h-5M921,0h5\" transform=\"translate(0,210.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M176,0h-5M921,0h5\" transform=\"translate(0,227.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M176,0h-5M921,0h5\" transform=\"translate(0,244.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M176,0h-5M921,0h5\" transform=\"translate(0,261.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><g class=\"ytick\"><text text-anchor=\"end\" x=\"168.6\" y=\"4.199999999999999\" transform=\"translate(0,108.5)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\">5:LYZ FCN1 CD14</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"168.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,125.5)\">10:BCL11B SKAP1 PRKCH</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"168.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,142.5)\">4:IL7R RPSA RPS5</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"168.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,159.5)\">7:ZEB2 AOAH TYROBP</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"168.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,176.5)\">2:NKG7 GNLY GZMA</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"168.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,193.5)\">3:CCL5 GZMK ATP8A1</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"168.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,210.5)\">1:MS4A1 CD79A BANK1</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"168.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,227.5)\">9:LTB BACH2 CD69</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"168.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,244.5)\">8:LYN CD83 HLA-DRA</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"168.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,261.5)\">6:CST3 PLXDC2 LRMDA</text></g></g><g class=\"overaxes-above\"/></g></g><g class=\"polarlayer\"/><g class=\"smithlayer\"/><g class=\"ternarylayer\"/><g class=\"geolayer\"/><g class=\"funnelarealayer\"/><g class=\"pielayer\"/><g class=\"iciclelayer\"/><g class=\"treemaplayer\"/><g class=\"sunburstlayer\"/><g class=\"glimages\"/><defs id=\"topdefs-ad8c6f\"><g class=\"clips\"/></defs><g class=\"layer-above\"><g class=\"imagelayer\"/><g class=\"shapelayer\"/></g><g class=\"infolayer\"><g class=\"g-gtitle\"><text class=\"gtitle\" x=\"500\" y=\"50\" text-anchor=\"middle\" dy=\"0.7em\" style=\"opacity: 1; font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 22px; fill: rgb(0, 0, 0); fill-opacity: 1; white-space: pre;\"><tspan style=\"font-weight:bold\">Hierarchical Clustering</tspan></text></g><g class=\"g-xtitle\"/><g class=\"g-ytitle\"/></g></svg>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import plotly.io as pio\n",
    "pio.renderers.default = \"svg\"\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:lda:n_documents: 9910\n",
      "INFO:lda:vocab_size: 1531\n",
      "INFO:lda:n_words: 308277\n",
      "INFO:lda:n_topics: 10\n",
      "INFO:lda:n_iter: 200\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selecting the Anchored LDA model.\n",
      "Number of topics: 10\n",
      "n_topics=10, n_iter=200, alpha=0.1,eta=0.1,random_state=9, refresh=30\n",
      "seed_topics={1412: 0, 1087: 0, 283: 1, 1425: 1, 1322: 2, 640: 2, 632: 3, 579: 3, 1270: 3, 686: 3, 694: 4, 152: 4, 1180: 4, 1086: 4, 1449: 5, 145: 5}, seed_confidence=0.15\n",
      "deconv_df.shape=(9910, 1531)\n",
      "(9910, 1531)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:lda:all zero column in document-term matrix found\n",
      "INFO:lda:<0> log likelihood: -1999309\n",
      "INFO:lda:<30> log likelihood: -1200556\n",
      "INFO:lda:<60> log likelihood: -1181604\n",
      "INFO:lda:<90> log likelihood: -1179109\n",
      "INFO:lda:<120> log likelihood: -1175993\n",
      "INFO:lda:<150> log likelihood: -1174819\n",
      "INFO:lda:<180> log likelihood: -1174626\n",
      "INFO:lda:<199> log likelihood: -1173867\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "WARNING:matplotlib.font_manager:findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n"
     ]
    }
   ],
   "source": [
    "sct.trainLDA(seed=9, n_iter=200,n_ensemble=0,ntopics=ntopics,\n",
    "            alpha=None,eta=None, # 默认为1/ntopics\n",
    "            refresh=30,seed_confidence=0.15,ll_plot=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively, hierarchical clustering can be applied to visualize the relationships among topics generated by the STED-LDA model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg class=\"main-svg\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"1000\" height=\"350\" style=\"\" viewBox=\"0 0 1000 350\"><rect x=\"0\" y=\"0\" width=\"1000\" height=\"350\" style=\"fill: rgb(255, 255, 255); fill-opacity: 1;\"/><defs id=\"defs-536367\"><g class=\"clips\"><clipPath id=\"clip536367xyplot\" class=\"plotclip\"><rect width=\"698\" height=\"170\"/></clipPath><clipPath class=\"axesclip\" id=\"clip536367x\"><rect x=\"222\" y=\"0\" width=\"698\" height=\"350\"/></clipPath><clipPath class=\"axesclip\" id=\"clip536367y\"><rect x=\"0\" y=\"100\" width=\"1000\" height=\"170\"/></clipPath><clipPath class=\"axesclip\" id=\"clip536367xy\"><rect x=\"222\" y=\"100\" width=\"698\" height=\"170\"/></clipPath></g><g class=\"gradients\"/><g class=\"patterns\"/></defs><g class=\"bglayer\"><rect class=\"bg\" x=\"222\" y=\"100\" width=\"698\" height=\"170\" style=\"fill: rgb(236, 239, 241); fill-opacity: 1; stroke-width: 0;\"/></g><g class=\"layer-below\"><g class=\"imagelayer\"/><g class=\"shapelayer\"/></g><g class=\"cartesianlayer\"><g class=\"subplot xy\"><g class=\"layer-subplot\"><g class=\"shapelayer\"/><g class=\"imagelayer\"/></g><g class=\"minor-gridlayer\"><g class=\"x\"/><g class=\"y\"/></g><g class=\"gridlayer\"><g class=\"x\"/><g class=\"y\"/></g><g class=\"zerolinelayer\"/><g class=\"layer-between\"><g class=\"shapelayer\"/><g class=\"imagelayer\"/></g><path class=\"xlines-below\"/><path class=\"ylines-below\"/><g class=\"overlines-below\"/><g class=\"xaxislayer-below\"/><g class=\"yaxislayer-below\"/><g class=\"overaxes-below\"/><g class=\"overplot\"><g class=\"xy\" transform=\"translate(222,100)\" clip-path=\"url(#clip536367xyplot)\"><g class=\"scatterlayer mlayer\"><g class=\"trace scatter trace34d298\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M0,59.5L154.48,59.5L154.48,76.5L0,76.5\" style=\"vector-effect: none; fill: none; stroke: rgb(61, 153, 112); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter traced78db6\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M154.48,68L188.86,68L188.86,93.5L0,93.5\" style=\"vector-effect: none; fill: none; stroke: rgb(61, 153, 112); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter traced7c33c\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M0,42.5L371.75,42.5L371.75,80.75L188.86,80.75\" style=\"vector-effect: none; fill: none; stroke: rgb(61, 153, 112); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter trace8cb17f\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M371.75,61.63L459.49,61.63L459.49,110.5L0,110.5\" style=\"vector-effect: none; fill: none; stroke: rgb(61, 153, 112); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter trace56edc4\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M0,25.5L473.04,25.5L473.04,86.06L459.49,86.06\" style=\"vector-effect: none; fill: none; stroke: rgb(61, 153, 112); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter trace037162\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M473.04,55.78L521.95,55.78L521.95,127.5L0,127.5\" style=\"vector-effect: none; fill: none; stroke: rgb(61, 153, 112); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter trace344852\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M0,8.5L572.72,8.5L572.72,91.64L521.95,91.64\" style=\"vector-effect: none; fill: none; stroke: rgb(61, 153, 112); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter trace706988\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M0,144.5L613.87,144.5L613.87,161.5L0,161.5\" style=\"vector-effect: none; fill: none; stroke: rgb(61, 153, 112); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g><g class=\"trace scatter trace749d18\" style=\"stroke-miterlimit: 2; opacity: 1;\"><g class=\"fills\"/><g class=\"errorbars\"/><g class=\"lines\"><path class=\"js-line\" d=\"M572.72,50.07L698,50.07L698,153L613.87,153\" style=\"vector-effect: none; fill: none; stroke: rgb(61, 153, 112); stroke-opacity: 1; stroke-width: 2px; opacity: 1;\"/></g><g class=\"points\"/><g class=\"text\"/></g></g></g></g><path class=\"xlines-above crisp\" d=\"M221,270.5H921M221,99.5H921\" style=\"fill: none; stroke-width: 1px; stroke: rgb(235, 240, 248); stroke-opacity: 1;\"/><path class=\"ylines-above crisp\" d=\"M221.5,100V270M920.5,100V270\" style=\"fill: none; stroke-width: 1px; stroke: rgb(235, 240, 248); stroke-opacity: 1;\"/><g class=\"overlines-above\"/><g class=\"xaxislayer-above\"><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(222,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(296.12,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(370.25,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(444.37,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(518.49,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(592.61,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(666.74,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(740.86,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(814.98,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"xtick ticks crisp\" d=\"M0,271v5M0,99v-5\" transform=\"translate(889.11,0)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" transform=\"translate(222,0)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\">0</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(296.12,0)\">0.1</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(370.25,0)\">0.2</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(444.37,0)\">0.3</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(518.49,0)\">0.4</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(592.61,0)\">0.5</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(666.74,0)\">0.6</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(740.86,0)\">0.7</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(814.98,0)\">0.8</text></g><g class=\"xtick\"><text text-anchor=\"middle\" x=\"0\" y=\"290.4\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(889.11,0)\">0.9</text></g></g><g class=\"yaxislayer-above\"><path class=\"ytick ticks crisp\" d=\"M221,0h-5M921,0h5\" transform=\"translate(0,108.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M221,0h-5M921,0h5\" transform=\"translate(0,125.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M221,0h-5M921,0h5\" transform=\"translate(0,142.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M221,0h-5M921,0h5\" transform=\"translate(0,159.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M221,0h-5M921,0h5\" transform=\"translate(0,176.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M221,0h-5M921,0h5\" transform=\"translate(0,193.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M221,0h-5M921,0h5\" transform=\"translate(0,210.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M221,0h-5M921,0h5\" transform=\"translate(0,227.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M221,0h-5M921,0h5\" transform=\"translate(0,244.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><path class=\"ytick ticks crisp\" d=\"M221,0h-5M921,0h5\" transform=\"translate(0,261.5)\" style=\"stroke: rgb(68, 68, 68); stroke-opacity: 1; stroke-width: 1px;\"/><g class=\"ytick\"><text text-anchor=\"end\" x=\"213.6\" y=\"4.199999999999999\" transform=\"translate(0,108.5)\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\">6:AFF3 BANK1 RALGPS2</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"213.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,125.5)\">1:CCL4 NKG7 GNLY</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"213.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,142.5)\">7:IFNG-AS1 INPP4B AC079793.1</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"213.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,159.5)\">8:TPD52 CD1D AC099063.4</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"213.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,176.5)\">2:SLC4A10 AC007368.1 PIP5K1B</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"213.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,193.5)\">0:NELL2 PAWR SAP30</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"213.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,210.5)\">5:FCGR3A CST3 TCF7L2</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"213.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,227.5)\">9:RPS15 RPS5 RAB27B</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"213.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,244.5)\">3:HIST1H4C SYNE1 DIP2A</text></g><g class=\"ytick\"><text text-anchor=\"end\" x=\"213.6\" y=\"4.199999999999999\" style=\"font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 12px; fill: rgb(42, 63, 95); fill-opacity: 1; white-space: pre; opacity: 1;\" transform=\"translate(0,261.5)\">4:VCAN CXCL8 PLCB1</text></g></g><g class=\"overaxes-above\"/></g></g><g class=\"polarlayer\"/><g class=\"smithlayer\"/><g class=\"ternarylayer\"/><g class=\"geolayer\"/><g class=\"funnelarealayer\"/><g class=\"pielayer\"/><g class=\"iciclelayer\"/><g class=\"treemaplayer\"/><g class=\"sunburstlayer\"/><g class=\"glimages\"/><defs id=\"topdefs-536367\"><g class=\"clips\"/></defs><g class=\"layer-above\"><g class=\"imagelayer\"/><g class=\"shapelayer\"/></g><g class=\"infolayer\"><g class=\"g-gtitle\"><text class=\"gtitle\" x=\"500\" y=\"50\" text-anchor=\"middle\" dy=\"0.7em\" style=\"opacity: 1; font-family: 'Open Sans', verdana, arial, sans-serif; font-size: 22px; fill: rgb(0, 0, 0); fill-opacity: 1; white-space: pre;\"><tspan style=\"font-weight:bold\">Hierarchical Clustering</tspan></text></g><g class=\"g-xtitle\"/><g class=\"g-ytitle\"/></g></svg>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import plotly.io as pio\n",
    "import plotly.graph_objs as go\n",
    "pio.renderers.default = \"svg\"\n",
    "fig = sct.model.plot_hierarchy()\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Annotate single-cell topic models biologically (GO enrichment) and generate plots.\n",
    "Since a topic often contains several gene sub-clusters, **TopicAnnotation** selects the most significant ones and maps the topic to specific cell types using the Cell Type Probability matrix. It then extracts high-weight genes for each topic to perform enrichment analysis and visualization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pkg_resources\n",
    "\n",
    "r_script_path = pkg_resources.resource_filename('STED', 'R/TopicAnnotation.R')\n",
    "cmd = f'Rscript \"{r_script_path}\" -i \"{out_dir}\" -g human -s \"{seed}\"'\n",
    "os.system(cmd)\n",
    "\n",
    "# or just run\n",
    "# !Rscript ./R/TopicAnnotation.R -i \"{out_dir}\" -g human -s \"{seed}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Bulk File Preparation and Preprocessing #"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = './demo/PBMC'\n",
    "out_dir = './demo/PBMC/test_CorEx'\n",
    "gs_count_file = os.path.join(data_dir,\"peaks_GAS.txt\")\n",
    "peak_count_file = os.path.join(data_dir,\"peaks_counts.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, preprocess the single-cell data as before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "scp = scPreProcessing()\n",
    "scp.set_data(sc_count_file=sc_count_file,sc_anno_file=sc_anno_file)\n",
    "scp.cell_selection()\n",
    "gene_use = pd.read_csv(os.path.join(data_dir,\"use_gene.csv\"),index_col=0)['0'].to_list()\n",
    "scp.geneWords(sc_anno_file, gene_use=gene_use)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "epid = gsPreProcessing()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "The **gs_count_file** can be either bulk RNA-Seq count data or Gene Activity Scores (GAS) derived from epigenomic data.<br>\n",
    "<br>\n",
    "If using epigenomic data, please set **Epi=True**.<br>\n",
    "When Epi=True, if gs_count_file is set to None by default, the gsPreProcessing function will automatically calculate the **Gene Activity Scores (GAS)** based on the peak_count_file. Alternatively, you may input a pre-calculated GAS file, as the GAS computation can be time-consuming.<br>\n",
    "<br>\n",
    "The **peak_count_file** is a read count file for peaks generated by tools such as featureCounts, supporting both BED and H5AD formats.<br>\n",
    "<br>\n",
    "Since **gene annotation (gene_anno_file)** is required for GAS calculation, it defaults to the hg38 genome assembly if set to None. For other species, please provide a corresponding gene annotation file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Epigenomic Deconvolution Input\n",
    "# epid.set_data(Epi=True,peak_file=peak_count_file,gene_anno_file=gene_anno_file) #自动计算GAS\n",
    "epid.set_data(Epi=True,peak_file=peak_count_file,gs_count_file=gs_count_file) #输入计算好的GAS\n",
    "# RNA Deconvolution Input\n",
    "# epid.set_data(Epi=False,gs_count_file=gs_count_file) #输入bulk RNA-seq counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Only calculate for genes present in scTopic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "epid.gene_selection(scp.use_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count_per_cell shape: (1,)\n"
     ]
    }
   ],
   "source": [
    "# Parameter Configuration for Data Preprocessing using the STED-CorEx Model\n",
    "epid.preprocessing(linear2log=True)\n",
    "\n",
    "# Parameter Configuration for Data Preprocessing using the STED-LDA Model\n",
    "# epid.preprocessing(linear2log=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: Filter out genes that have no signal in bulk samples before performing deconvolution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "scp.geneShared(epid.gs_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Subsequently, construct the scTopic object for model training."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- reflect genes in expression ---\n",
      "6 anchored topics\n",
      "16 genes were registered\n",
      "--- extract cell type specific anchors ---\n",
      "6 cells\n",
      "set() type of anchored genes were removed (no specific after removing overlap)\n",
      "16 genes were registered\n",
      "seed number: 16\n",
      "seed_k: 16\n",
      "done.\n",
      "Selecting the Anchored CorEx model.\n",
      "WARNING: Some words never appear (or always appear)\n",
      "WARNING: Some words never appear (or always appear)\n",
      "weight threshold is 0.000000 for graph with max of 102.000000 edges \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n",
      "findfont: Generic family 'sans-serif' not found because none of the following families were found: Arial\n"
     ]
    }
   ],
   "source": [
    "scp.preprocessing(linear2log=True)\n",
    "scp.preForGuide(anchored_genes=anchor_words)\n",
    "\n",
    "sct = scTopic()\n",
    "sct.setData(scp,out_dir)\n",
    "# 如果重点是做反卷积任务，我们推荐设置高一点的topic number，此处由于PBMC细胞类型少，只有6种，我们以15例子\n",
    "ntopics = 15\n",
    "sct.trainCorEx(seed=seed,n_iter=200,n_ensemble=0,ntopics=ntopics,\n",
    "               anchor_strength=2,hierarchy_topic = [6,2],tc_plot=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, create a scTopic object to be used for model training."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Cell proportion and epigenetic signal estimation via scTopic + epiDecon #"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Construct an epiDecon object for deconvolution and signal prediction. The out_dir parameter specifies the output directory for saving results and requires gsPreProcessing objects for both the preprocessed single-cell data and bulk data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "epi = epiDecon()\n",
    "epi.SetData(out_dir,scp,epid)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The model parameter accepts a pre-trained scTopic model. If no model is provided (model=None), epiDecon will automatically search for the necessary files within the specified out_dir path.<br>\n",
    "<br>\n",
    "Setting **benchmark=True** instructs epiDecon to save intermediate files generated during the deconvolution process.\n",
    "If a model is specified, epiDecon will automatically adopt the optimal normalization method selected during the scTopic training phase. <br>\n",
    "<br>\n",
    "Conversely, if model=None, you must manually specify the normalization method for the topic-celltype correlation matrix. Supported methods include: \"Raw\", \"Norm\", \"NormBySD\", \"Bayes\", and \"BayesNorm\" (default)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Raw**: Utilizes the original topic_celltype weights directly for calculation without any transformation.<br>\n",
    "<br>\n",
    "**Norm**: Applies column-wise sum normalization (Sum-Norm) to the original weights. This normalizes the weights of cell types under each topic to sum to one, effectively treating them as a probability distribution (i.e., the probability of being a specific cell type given that topic).<br>\n",
    "<br>\n",
    "**NormBySD**: Scales the weights using StandardScaler (with_mean=False). This rescales the weights by the standard deviation to eliminate differences in numerical magnitude across topics, without removing the mean.<br>\n",
    "<br>\n",
    "**Bayes**: Uses the celltype_topic_mat_bayes.txt matrix file generated via the Bayesian method, with no additional processing applied.<br>\n",
    "<br>\n",
    "**BayesNorm**: Uses the celltype_topic_mat_bayes.txt matrix file generated via the Bayesian method, followed by column normalization (ensuring the weights for each topic sum to one)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Some words never appear (or always appear)\n"
     ]
    }
   ],
   "source": [
    "epi.Decon(ntopics_selection=ntopics,seed_selection=seed,\n",
    "          model=sct,model_selection=\"CorEx\",#\"LDA\",\"ALDA\"(guided LDA)\n",
    "          outname=None,benchmark=False)\n",
    "epi.Bayes()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The predicted cell proportions are stored in the celltype_frac_df attribute of the gsPreProcessing object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>B cells</th>\n",
       "      <th>CD4 T cells</th>\n",
       "      <th>CD8 T cells</th>\n",
       "      <th>Dendritic cells</th>\n",
       "      <th>Monocytes</th>\n",
       "      <th>NK cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Bulk</th>\n",
       "      <td>0.044902</td>\n",
       "      <td>0.440934</td>\n",
       "      <td>0.198553</td>\n",
       "      <td>0.016285</td>\n",
       "      <td>0.245627</td>\n",
       "      <td>0.053699</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       B cells  CD4 T cells  CD8 T cells  Dendritic cells  Monocytes  NK cells\n",
       "Bulk  0.044902     0.440934     0.198553         0.016285   0.245627  0.053699"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epi.celltype_frac_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you need to predict epigenetic signals for each cell type, run the **PeakPredict** function, inputting the single-cell data and the bulk sample preprocessing object.<br>\n",
    "It supports directly reading the **gene2peak_project_mat** saved during GAS calculation.<br>\n",
    "If **GS_file** is set to None, epiDecon will automatically recalculate it; in this case, the gene_anno_file is required (can be ignored for human samples)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "GS_file = os.path.join(data_dir,\"gene2peak_project_mat.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lyx/.conda/envs/STED/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning:\n",
      "\n",
      "Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read gene peak weight matrix\n",
      "default peak name\n"
     ]
    }
   ],
   "source": [
    "predicted = epi.PeakPredict(sc_count_file,\n",
    "                            epid,\n",
    "                            GS_file=GS_file,\n",
    "                            gene_anno_file=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr1_865422_866353</th>\n",
       "      <th>chr1_869477_870378</th>\n",
       "      <th>chr1_876272_877037</th>\n",
       "      <th>chr1_877185_877929</th>\n",
       "      <th>chr1_896976_897832</th>\n",
       "      <th>chr1_898268_899165</th>\n",
       "      <th>chr1_904364_905213</th>\n",
       "      <th>chr1_910075_910856</th>\n",
       "      <th>chr1_920762_921638</th>\n",
       "      <th>chr1_923415_924300</th>\n",
       "      <th>...</th>\n",
       "      <th>chrX_154457667_154458606</th>\n",
       "      <th>chrX_154478633_154479530</th>\n",
       "      <th>chrX_154484222_154485044</th>\n",
       "      <th>chrX_154486250_154487107</th>\n",
       "      <th>chrX_154490305_154491211</th>\n",
       "      <th>chrX_154497366_154498281</th>\n",
       "      <th>chrX_155242134_155243020</th>\n",
       "      <th>chrX_155262680_155263367</th>\n",
       "      <th>chrX_155263990_155264871</th>\n",
       "      <th>chrX_155334535_155335348</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NK cells</th>\n",
       "      <td>16.081331</td>\n",
       "      <td>399.625369</td>\n",
       "      <td>9.115621</td>\n",
       "      <td>9.287613</td>\n",
       "      <td>10.147578</td>\n",
       "      <td>23.133037</td>\n",
       "      <td>327.560368</td>\n",
       "      <td>24.680973</td>\n",
       "      <td>81.782597</td>\n",
       "      <td>170.874889</td>\n",
       "      <td>...</td>\n",
       "      <td>244.600655</td>\n",
       "      <td>376.397902</td>\n",
       "      <td>12.475953</td>\n",
       "      <td>171.184470</td>\n",
       "      <td>332.492145</td>\n",
       "      <td>7.197665</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4 T cells</th>\n",
       "      <td>60.837581</td>\n",
       "      <td>1511.830164</td>\n",
       "      <td>34.485474</td>\n",
       "      <td>35.136143</td>\n",
       "      <td>38.389490</td>\n",
       "      <td>87.515023</td>\n",
       "      <td>1239.199719</td>\n",
       "      <td>93.371047</td>\n",
       "      <td>309.393261</td>\n",
       "      <td>646.439969</td>\n",
       "      <td>...</td>\n",
       "      <td>3541.783468</td>\n",
       "      <td>5450.189274</td>\n",
       "      <td>180.650064</td>\n",
       "      <td>2478.727322</td>\n",
       "      <td>4814.440010</td>\n",
       "      <td>104.221191</td>\n",
       "      <td>15.461649</td>\n",
       "      <td>5.322426</td>\n",
       "      <td>55.592745</td>\n",
       "      <td>8.223149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Monocytes</th>\n",
       "      <td>70.375059</td>\n",
       "      <td>1748.839024</td>\n",
       "      <td>39.891744</td>\n",
       "      <td>40.644419</td>\n",
       "      <td>44.407791</td>\n",
       "      <td>101.234710</td>\n",
       "      <td>1433.468440</td>\n",
       "      <td>108.008780</td>\n",
       "      <td>357.896688</td>\n",
       "      <td>747.782040</td>\n",
       "      <td>...</td>\n",
       "      <td>923.317229</td>\n",
       "      <td>1420.824763</td>\n",
       "      <td>47.094160</td>\n",
       "      <td>646.186212</td>\n",
       "      <td>1255.089547</td>\n",
       "      <td>27.169708</td>\n",
       "      <td>122.744066</td>\n",
       "      <td>42.252691</td>\n",
       "      <td>441.329353</td>\n",
       "      <td>65.280407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD8 T cells</th>\n",
       "      <td>47.829311</td>\n",
       "      <td>1188.571176</td>\n",
       "      <td>27.111802</td>\n",
       "      <td>27.623346</td>\n",
       "      <td>30.181063</td>\n",
       "      <td>68.802592</td>\n",
       "      <td>974.234476</td>\n",
       "      <td>73.406483</td>\n",
       "      <td>243.238904</td>\n",
       "      <td>508.218405</td>\n",
       "      <td>...</td>\n",
       "      <td>1406.360462</td>\n",
       "      <td>2164.144357</td>\n",
       "      <td>71.731971</td>\n",
       "      <td>984.245405</td>\n",
       "      <td>1911.702999</td>\n",
       "      <td>41.383829</td>\n",
       "      <td>15.757317</td>\n",
       "      <td>5.424206</td>\n",
       "      <td>56.655827</td>\n",
       "      <td>8.380398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B cells</th>\n",
       "      <td>3.367706</td>\n",
       "      <td>83.688398</td>\n",
       "      <td>1.908967</td>\n",
       "      <td>1.944985</td>\n",
       "      <td>2.125077</td>\n",
       "      <td>4.844454</td>\n",
       "      <td>68.596753</td>\n",
       "      <td>5.168619</td>\n",
       "      <td>17.126677</td>\n",
       "      <td>35.784129</td>\n",
       "      <td>...</td>\n",
       "      <td>269.406970</td>\n",
       "      <td>414.570510</td>\n",
       "      <td>13.741209</td>\n",
       "      <td>188.545241</td>\n",
       "      <td>366.212025</td>\n",
       "      <td>7.927621</td>\n",
       "      <td>4.992693</td>\n",
       "      <td>1.718655</td>\n",
       "      <td>17.951352</td>\n",
       "      <td>2.655322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dendritic cells</th>\n",
       "      <td>1.337845</td>\n",
       "      <td>33.245806</td>\n",
       "      <td>0.758351</td>\n",
       "      <td>0.772659</td>\n",
       "      <td>0.844202</td>\n",
       "      <td>1.924494</td>\n",
       "      <td>27.250544</td>\n",
       "      <td>2.053270</td>\n",
       "      <td>6.803693</td>\n",
       "      <td>14.215498</td>\n",
       "      <td>...</td>\n",
       "      <td>118.431325</td>\n",
       "      <td>182.245228</td>\n",
       "      <td>6.040636</td>\n",
       "      <td>82.884502</td>\n",
       "      <td>160.986834</td>\n",
       "      <td>3.484983</td>\n",
       "      <td>44.956669</td>\n",
       "      <td>15.475618</td>\n",
       "      <td>161.642825</td>\n",
       "      <td>23.909829</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 22697 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 chr1_865422_866353  chr1_869477_870378  chr1_876272_877037  \\\n",
       "NK cells                  16.081331          399.625369            9.115621   \n",
       "CD4 T cells               60.837581         1511.830164           34.485474   \n",
       "Monocytes                 70.375059         1748.839024           39.891744   \n",
       "CD8 T cells               47.829311         1188.571176           27.111802   \n",
       "B cells                    3.367706           83.688398            1.908967   \n",
       "Dendritic cells            1.337845           33.245806            0.758351   \n",
       "\n",
       "                 chr1_877185_877929  chr1_896976_897832  chr1_898268_899165  \\\n",
       "NK cells                   9.287613           10.147578           23.133037   \n",
       "CD4 T cells               35.136143           38.389490           87.515023   \n",
       "Monocytes                 40.644419           44.407791          101.234710   \n",
       "CD8 T cells               27.623346           30.181063           68.802592   \n",
       "B cells                    1.944985            2.125077            4.844454   \n",
       "Dendritic cells            0.772659            0.844202            1.924494   \n",
       "\n",
       "                 chr1_904364_905213  chr1_910075_910856  chr1_920762_921638  \\\n",
       "NK cells                 327.560368           24.680973           81.782597   \n",
       "CD4 T cells             1239.199719           93.371047          309.393261   \n",
       "Monocytes               1433.468440          108.008780          357.896688   \n",
       "CD8 T cells              974.234476           73.406483          243.238904   \n",
       "B cells                   68.596753            5.168619           17.126677   \n",
       "Dendritic cells           27.250544            2.053270            6.803693   \n",
       "\n",
       "                 chr1_923415_924300  ...  chrX_154457667_154458606  \\\n",
       "NK cells                 170.874889  ...                244.600655   \n",
       "CD4 T cells              646.439969  ...               3541.783468   \n",
       "Monocytes                747.782040  ...                923.317229   \n",
       "CD8 T cells              508.218405  ...               1406.360462   \n",
       "B cells                   35.784129  ...                269.406970   \n",
       "Dendritic cells           14.215498  ...                118.431325   \n",
       "\n",
       "                 chrX_154478633_154479530  chrX_154484222_154485044  \\\n",
       "NK cells                       376.397902                 12.475953   \n",
       "CD4 T cells                   5450.189274                180.650064   \n",
       "Monocytes                     1420.824763                 47.094160   \n",
       "CD8 T cells                   2164.144357                 71.731971   \n",
       "B cells                        414.570510                 13.741209   \n",
       "Dendritic cells                182.245228                  6.040636   \n",
       "\n",
       "                 chrX_154486250_154487107  chrX_154490305_154491211  \\\n",
       "NK cells                       171.184470                332.492145   \n",
       "CD4 T cells                   2478.727322               4814.440010   \n",
       "Monocytes                      646.186212               1255.089547   \n",
       "CD8 T cells                    984.245405               1911.702999   \n",
       "B cells                        188.545241                366.212025   \n",
       "Dendritic cells                 82.884502                160.986834   \n",
       "\n",
       "                 chrX_154497366_154498281  chrX_155242134_155243020  \\\n",
       "NK cells                         7.197665                  0.000000   \n",
       "CD4 T cells                    104.221191                 15.461649   \n",
       "Monocytes                       27.169708                122.744066   \n",
       "CD8 T cells                     41.383829                 15.757317   \n",
       "B cells                          7.927621                  4.992693   \n",
       "Dendritic cells                  3.484983                 44.956669   \n",
       "\n",
       "                 chrX_155262680_155263367  chrX_155263990_155264871  \\\n",
       "NK cells                         0.000000                  0.000000   \n",
       "CD4 T cells                      5.322426                 55.592745   \n",
       "Monocytes                       42.252691                441.329353   \n",
       "CD8 T cells                      5.424206                 56.655827   \n",
       "B cells                          1.718655                 17.951352   \n",
       "Dendritic cells                 15.475618                161.642825   \n",
       "\n",
       "                 chrX_155334535_155335348  \n",
       "NK cells                         0.000000  \n",
       "CD4 T cells                      8.223149  \n",
       "Monocytes                       65.280407  \n",
       "CD8 T cells                      8.380398  \n",
       "B cells                          2.655322  \n",
       "Dendritic cells                 23.909829  \n",
       "\n",
       "[6 rows x 22697 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predicted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "STED",
   "language": "python",
   "name": "sted"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
